transport/session/src/types.rs

```diff
@@ -0,0 +1,357 @@
+use std::{
+    io::{Error, ErrorKind},
+    task::Poll,
+};
+
+use futures::{channel::mpsc::UnboundedReceiver, FutureExt, StreamExt};
+use libp2p_identity::PeerId;
+
+use hopr_crypto_types::types::OffchainPublicKey;
+use hopr_internal_types::protocol::{ApplicationData, PAYLOAD_SIZE};
+
+use crate::{errors::TransportSessionError, traits::SendMsg, PathOptions};
+
+/// ID tracking the session uniquely.
+///
+/// Simple wrapper around the maximum range of the port like session unique identifier.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct SessionId {
+    tag: u16,
+    peer: PeerId,
+}
+
+impl SessionId {
+    pub fn new(tag: u16, peer: PeerId) -> Self {
+        Self { tag, peer }
+    }
+
+    pub fn tag(&self) -> u16 {
+        self.tag
+    }
+
+    pub fn peer(&self) -> &PeerId {
+        &self.peer
+    }
+}
+
+pub struct Session {
+    id: SessionId,
+    me: PeerId,
+    options: PathOptions,
+    rx: UnboundedReceiver<Box<[u8]>>,
+    tx: Box<dyn SendMsg>,
+    rx_buffer: [u8; PAYLOAD_SIZE],
+    rx_buffer_range: (usize, usize),
+}
+
+impl Session {
+    pub fn new(
+        id: SessionId,
+        me: PeerId,
+        options: PathOptions,
+        tx: Box<dyn SendMsg>,
+        rx: UnboundedReceiver<Box<[u8]>>,
+    ) -> Self {
+        Self {
+            id,
+            me,
+            options,
+            rx,
+            tx,
+            rx_buffer: [0; PAYLOAD_SIZE],
+            rx_buffer_range: (0, 0),
+        }
+    }
+
+    pub fn id(&self) -> &SessionId {
+        &self.id
+    }
+}
+
+impl futures::AsyncWrite for Session {
+    fn poll_write(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> Poll<std::io::Result<usize>> {
+        let payload = wrap_with_offchain_key(&self.me, buf.to_vec().into_boxed_slice())
+            .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
+
+        let payload = ApplicationData::new_from_owned(Some(self.id.tag()), payload.into_boxed_slice())
+            .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
+
+        match self
+            .tx
+            .send_message(payload, *self.id.peer(), self.options.clone())
+            .poll_unpin(cx)
+        {
+            Poll::Ready(Ok(_)) => Poll::Ready(Ok(buf.len())),
+            Poll::Ready(Err(_)) => Poll::Ready(Err(Error::from(ErrorKind::BrokenPipe))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+
+    fn poll_flush(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<std::io::Result<()>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<std::io::Result<()>> {
+        Poll::Ready(Ok(()))
```

Comment:
Should closing the session also close the receiver?

Reply:
I don't know.

Currently there's no mechanism for back pressure or other form of logical connection feedback. We will have to agree on a reasonable mechanism, but since the peer is `n-hops` away, there's no physical connection mechanism to be able to do "close" operations. It'd be possible to abstract the "closing" operation, but that would either:
1. be equal to discarding the `PathOptions` and the `Session` object
2. use mechanism like 0.hop ping to verify that the peer is still online

Reply:
But this will be sorted out once the session is being integrated.